implementations/rust/ockam/ockam_app/src/app/system_tray.rs

```diff
@@ -0,0 +1,70 @@
+use crate::ctx::TauriCtx;
+use crate::enroll::EnrollActions;
+use crate::quit::QuitActions;
+use crate::tcp::outlet::TcpOutletActions;
+use crate::Result;
+use tauri::{CustomMenuItem, SystemTrayMenu, SystemTrayMenuItem};
+
+/// Create the system tray with all the major functions.
+/// Separate groups of related functions with a native separator.
+pub struct SystemTrayMenuBuilder {
+    enroll: EnrollActions,
+    tcp: TcpOutletActions,
+    quit: QuitActions,
+}
+
+impl SystemTrayMenuBuilder {
+    /// Create the default system tray menu with the basic elements (i.e. without list items).
+    pub fn default() -> SystemTrayMenu {
+        Self::init().build()
+    }
+
+    pub fn init() -> Self {
+        let enroll = EnrollActions::new();
+        let tcp = TcpOutletActions::new();
+        let quit = QuitActions::new();
+        Self { enroll, tcp, quit }
+    }
+
+    /// Create a `SystemTrayMenu` instance, adding the menu items in the expected order.
+    pub fn build(self) -> SystemTrayMenu {
+        SystemTrayMenu::new()
+            .add_menu_items(&[self.enroll.enroll])
+            .add_native_item(SystemTrayMenuItem::Separator)
+            .add_menu_items(&self.tcp.menu_items)
+            .add_native_item(SystemTrayMenuItem::Separator)
+            .add_menu_items(&[self.enroll.reset, self.quit.quit])
+    }
+
+    /// Refresh the system tray menu with the latest state, including all list items.
+    pub fn refresh(ctx: &TauriCtx) -> Result<()> {
+        let menu = Self::get_full_menu(ctx).unwrap_or(Self::default());
+        ctx.app_handle().tray_handle().set_menu(menu)?;
+        Ok(())
+    }
+
+    fn get_full_menu(ctx: &TauriCtx) -> Result<SystemTrayMenu> {
+        let enroll = EnrollActions::new();
+        let tcp = TcpOutletActions::full(ctx)?;
```

Comment:
One improvement in the future would be to just gray out the items that can't be accessed in case of an error. Can you please create an issue for it?

Reply:
This call will fail if we fail to retrieve that information from the node. In that case, the TCP outlet list will be empty.

Reply:
My concern is that if anything changes in the implementation then `?` will short-circuit `get_full_menu` and nothing will be displayed.

Reply:
Note that if `get_full_menu` fails then the default menu (with basic menu items) is displayed:

https://github.com/build-trust/ockam/pull/5290/files#diff-344bd0bf09709196cb556acc10f998293791e3ac3a795b0c1dcbeccb43a58c31R41

Reply:
@adrianbenavides that's indeed at least a good backup. I was aiming for a more fine-grained behavior but there's no need to tackle this now.